### PR TITLE
Display metrics at the pipeline level

### PR DESCRIPTION
**Pull request by garg-amit**
## Purpose

* Log training metadata in MLFlow to showcase metrics at the pipeline level.
* Report train/test loss and test accuracy for each round.
* Display metrics for each as well as combined rounds corresponding to all silos.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

Submit a pipeline and then go to the AML UI to see metrics in the "Pipeline Job overview -> Metrics"

### CODE REVIEW

File: examples/components/MNIST/preprocessing/run.py [hunk 1]
```diff
@@ -117,13 +118,35 @@ def preprocess_data(
     test_dataset = MnistDataset(X_test.float(), y_test, transform=transform)
     datasets = {"train": train_dataset, "test": test_dataset}
 
+    # Mlflow logging
+    log_metdata(X_train, X_test)
```
Comment (jfomhover) [inline examples/components/MNIST/preprocessing/run.py:122]: typo: log_metadata

File: examples/components/MNIST/traininsilo/traininsilo.yaml [hunk 2]
```diff
@@ -14,6 +14,16 @@ inputs:
   test_data:
     type: uri_folder
     description: the input testing data
+  silo_name:
```
Comment (jfomhover) [inline examples/components/MNIST/traininsilo/traininsilo.yaml:17]: Could we instead use a non-FL name for this? for instance `metrics_prefix`. This way, it's still a generic component, but we use it in a way that fits in the FL scenario.

File: examples/components/MNIST/traininsilo/run.py [hunk 3]
```diff
@@ -92,22 +97,69 @@ def load_dataset(self, train_data_dir, test_data_dir):
...
+        client.log_param(
+            run_id=run_id,
+            key=f"optimizer {self._experiment_name}",
+            value=self.optimizer_.__class__.__name__,
+        )
+
+    def log_metrics(self, client, run_id, key, value):
+        client.log_metric(
+            run_id=run_id,
+            key=key + f", {self._experiment_name}",
```
Comment (jfomhover) [inline examples/components/MNIST/traininsilo/run.py:126]: could we use some programmatically easy naming convention?

for instance: {iteration}/{silo}/{metric} so that in the UI it becomes easy to browse visually, and we could also parse those in mlflow later?

**Review by jfomhover — CHANGES_REQUESTED** [review]

File: examples/pipelines/fl_cross_silo_basic/config.yaml [hunk 4]
```diff
@@ -24,7 +25,8 @@ federated_learning:
         type: uri_file
         mode: 'download'
         path: https://azureopendatastorage.blob.core.windows.net/mnist/processed/t10k.csv
-    - compute: cpu-cluster-westus
+    - name: Silo-2
```
Comment (jfomhover) [inline examples/pipelines/fl_cross_silo_basic/config.yaml:28]: could you use the region as a name? It would be more catchy in a demo to see which region is lighting up in the pipeline

File: examples/components/MNIST/preprocessing/run.py (hunk 1 above)
Comment (jfomhover) [reply examples/components/MNIST/preprocessing/run.py:122]: also, could we also use metrics_prefix here to display the metadata under a metric name that shows which silo it is?

**Review by jfomhover — CHANGES_REQUESTED** [review]

File: examples/pipelines/fl_cross_silo_basic/config.yaml (hunk 4 above)
Comment (garg-amit) [reply examples/pipelines/fl_cross_silo_basic/config.yaml]: Sure.

File: examples/components/MNIST/traininsilo/run.py (hunk 3 above)
Comment (garg-amit) [reply examples/components/MNIST/traininsilo/run.py]: Loved this

File: examples/components/MNIST/traininsilo/traininsilo.yaml (hunk 2 above)
Comment (garg-amit) [reply examples/components/MNIST/traininsilo/traininsilo.yaml]: oh makes sense.

File: examples/components/MNIST/preprocessing/run.py (hunk 1 above)
Comment (garg-amit) [reply examples/components/MNIST/preprocessing/run.py]: sure, did the required changes.

**Review by jfomhover — APPROVED** [review]
Looks good thanks!
